[MFT] Standalone track fitting

rpezzi opened the pull request:
This is the first version of MFT standalone track fitting. Fitting methods were adapted from MCH. The coordinate system has been adapted to the MFT: (X, Y, Phi, Tanl, q/pt). Further improvements are planned: refine track seed covariance matrix, add MS effects, create an alternative to TrackParCov class suitable for MFT tracks in this coordinate system; create a MUON base cluster class.

bovulpes commented:
It is ok for me to merge this PR.

shahor02 commented on Detectors/ITSMFT/MFT/tracking/include/MFTTracking/TrackParam.h:
```diff
@@ -0,0 +1,182 @@
+  /// PHI     = azimutal angle
+  /// TANL    = tangent of \lambda (dip angle)
+  /// INVQPT    = Inverse transverse momentum (GeV/c ** -1) times charge (assumed forward motion)  </pre>
+  TMatrixD mParameters{5, 1}; ///< \brief Track parameters
```
Why using TMatrixD, even for internal use? It derives from the TObject so has all the virtuality overheads. Also, TMatrixD is an arbitrary dim. matrix, while your dimensions are fixed. Why don't you use a std::array here (or, if you want to use matrix operations, SVector class)?
Also, do you really need double?

rpezzi replied:
See comment below.

shahor02 commented on Detectors/ITSMFT/MFT/tracking/include/MFTTracking/TrackParam.h:
```diff
@@ -0,0 +1,182 @@
+  /// <X,PHI>       <Y,PHI>         <PHI,PHI>     <TANL,PHI>      <INVQPT,PHI>
+  /// <X,TANL>      <Y,TANL>       <PHI,TANL>     <TANL,TANL>     <INVQPT,TANL>
+  /// <X,INVQPT>   <Y,INVQPT>     <PHI,INVQPT>   <TANL,INVQPT>   <INVQPT,INVQPT>  </pre>
+  mutable std::unique_ptr<TMatrixD> mCovariances{}; ///< \brief Covariance matrix of track parameters
```
The cov.matrix is symmetric, either use array of 15 values (preferable) or ``ROOT::Math::SMatrix<double, 5, 5, ROOT::Math::MatRepSym<double, 5>>;``, avoid TMatrixD at any rate!

rpezzi replied:
See comment bellow. SMatrix will most certainly be used.

shahor02 commented on Detectors/ITSMFT/MFT/tracking/src/TrackExtrap.cxx:
```diff
@@ -0,0 +1,276 @@
+  double tanl0sq = tanl0 * tanl0;
+  double k = -mBZField * 0.3;
+
+  TMatrixD jacob(5, 5);
```
idem, you can use here a fixed size matrix, no need for TMatrixD

rpezzi replied:
idem

ktf commented:
Please fix the code formatting.

rpezzi commented on Detectors/ITSMFT/MFT/tracking/include/MFTTracking/TrackExtrap.h:
```diff
@@ -0,0 +1,65 @@
+  TrackExtrap& operator=(TrackExtrap&&) = delete;
+
+  void setBz(float bZ) { mBZField = bZ; } /// Set the magnetic field for the MFT
+  float getBz() { return mBZField; }
```
Done

rpezzi commented on Detectors/ITSMFT/MFT/tracking/include/MFTTracking/TrackExtrap.h:
```diff
@@ -0,0 +1,65 @@
+  static void addMCSEffect(TrackParam* trackParam, double dZ, double x0, bool isFieldON = true);
+
+ private:
+  static Float_t mBZField; // Tesla.
```
Fixed

rpezzi commented on Detectors/ITSMFT/MFT/tracking/include/MFTTracking/TrackFitter.h:
```diff
@@ -0,0 +1,85 @@
+  /// Enable/disable the smoother (and the saving of related parameters)
+  void smoothTracks(bool smooth) { mSmooth = smooth; }
+  /// Return the smoother enable/disable flag
+  bool isSmootherEnabled() { return mSmooth; }
```
Fixed

rpezzi commented on Detectors/ITSMFT/MFT/tracking/include/MFTTracking/TrackFitter.h:
```diff
@@ -0,0 +1,85 @@
+  Float_t mBZField = 0.5;                   // Tesla.
+  static constexpr double SMaxChi2 = 2.e10; ///< maximum chi2 above which the track can be considered as abnormal
+  /// z position of the layers
+  static constexpr float SDefaultLayerZ[10] = {-45.3, -46.7, -48.6, -50.0, -52.4, -53.8, -67.7, -69.1, -76.1, -77.5};
```
Fixed. Now using o2::mft::constants from `MFTBase/Constants.h`

rpezzi commented:
Dear @sawenzel  and @shahor02 , thanks for your inputs. I have addressed most of your comments. Particularly relevant is the creation of a new track class to store the MFT tracks. This class is derived from the track fitting parameter class, which will be trimmed once we decide if this coordinate system is successful. For proper assessment of the fitting quality the MC labels need to be propagated (TODO). As for the use of TMatrixD, we are using the same methods as the MCH already in O2@dev. I rather postpone this optimization until we decide on a common approach.

shahor02 commented on DataFormats/Detectors/ITSMFT/MFT/include/DataFormatsMFT/TrackParamMFT.h:
```diff
@@ -0,0 +1,200 @@
+  Double_t mLocalChi2 = 0.; ///< Local chi2 of the associated cluster with respect to the track
+};
+
+class TrackMFTParam : public TrackParamMFT
```
deriving TrackMFTParam from TrackParamMFT is somewhat confusing, perhaps more appropriate names can be used? But I don't see  where the TrackMFTParam is used at all. Is this class needed?

shahor02 commented on DataFormats/Detectors/ITSMFT/MFT/src/TrackParamMFT.cxx:
```diff
@@ -0,0 +1,370 @@
+
+using namespace std;
+
+//_________________________________________________________________________
```
As far as I understand, the TrackParamMFT is supposed to be used only for internal, while the TrackMFT will get its own data members after the validation of new track model, right? 
Then, I think the TrackParamMFT should not be in the DataFormats directory since there we collect classes which are supposed to be messaged over the DPL. I.e. the TrackMFT should stay in the DataFormats and become asap the meseagable (POD) class, while the TrackParamMFT it is better to move to the MFT algorithms directory. Also, since at the moment it is a base class of the TrackMFT which must be persistent, I think you should add ClassDefNV to it.

rpezzi replied:
Yes, Ruben, you are right. The inheritance from TrackParamMFT was a dirty solution to get the track at the dataformat directory. I will follow your suggestion.

shahor02 replied:
If due to the dependencies you need to keep the TrackParamMFT in the DataFormats, we can keep it at the moment there until you detach them. Anyway, if you keep the inheritance, the TrackMFT will not be messageable.

rpezzi replied:
Hi @shahor02 , TrackMFT has been cleaned-up and TrackParamMFT is back to where it belongs on the MFT Tracking directory.

ktf commented on Detectors/ITSMFT/MFT/workflow/src/TrackerSpec.cxx:
```diff
@@ -129,21 +129,21 @@ void TrackerDPL::run(ProcessingContext& pc)
         mTracker->setROFrame(roFrame);
         mTracker->clustersToTracks(event);
-        tracks.swap(mTracker->getTracks());
+        //tracks.swap(mTracker->getTracks());
```
Please do not comment code. Remove if not needed.

rpezzi replied:
Hi @ktf , I reviewed the code and removed some commented code. I am keeping those used for debugging.

shahor02 commented on Detectors/ITSMFT/MFT/workflow/src/TrackFitterSpec.cxx:
```diff
@@ -0,0 +1,175 @@
+    return;
+
+  auto tracksLTF = pc.inputs().get<const std::vector<o2::mft::TrackLTF>>("tracksltf");
+  auto tracksCA = pc.inputs().get<const std::vector<o2::mft::TrackCA>>("tracksca");
```
both TrackCA and TrackLTF are POD classes, instead of ``get<const std::vector<...>>`` one should use ``get<gsl::span<...>>`` (one can do this in separate PR).

shahor02 commented on Detectors/ITSMFT/MFT/workflow/src/TrackerSpec.cxx:
```diff
@@ -92,10 +92,10 @@ void TrackerDPL::run(ProcessingContext& pc)
   }
 
-  std::vector<o2::mft::TrackMFTExt> tracks;
+  //std::vector<o2::mft::TrackMFTExt> tracks;
```
Also in this device the clusters are extracted (L73,74) via vectors (i.e. copy is created) instead of accessing them via span. 
Since this is out of scope of this PR, I propose that we merge it, but then please open another PR changing vectors to spans.
